src/vtl.ts

```diff
@@ -418,4 +437,78 @@ export class VTL {
     const __exhaustive: never = node;
     return __exhaustive;
   }
+
+  private mapForEachArgs(call: CallExpr) {
+    const fn = call.args.callbackfn as FunctionExpr;
+    return fn.parameters.map((p) => p.name);
+  }
+
+  private renderMapOrForEach(
+    call: CallExpr,
+    list: string,
+    // Should start with $
+    returnVariable?: string,
+    index?: string,
+    array?: string
+  ) {
+    if (index) {
+      this.add(`#set($${index} = $foreach.index)`);
+    }
+    if (array) {
+      this.add(`#set($${array} = ${list})`);
+    }
+
+    const fn = call.args.callbackfn as FunctionExpr;
+
+    const tmp = returnVariable ? returnVariable : this.newLocalVarName();
+
+    for (const stmt of fn.body.statements) {
+      this.eval(stmt, tmp);
+    }
+
+    return tmp;
+  }
+
+  /**
+   * Recursively dives into map operations until a non-map or a map with arr paremeter is found.
+   * evaluates the expression after the last map.
+   *
+   * @return [firstVariable, list variable, render function]
+   */
+  private divePropAccessMapOperations(
+    expr: Expr,
+    // Should start with $
+    returnVariable: string,
+    alwaysEvaluate?: boolean
+  ): [string, string, () => void] {
+    if (
+      !alwaysEvaluate &&
+      expr.kind === "CallExpr" &&
+      expr.expr.kind === "PropAccessExpr" &&
+      expr.expr.name === "map"
+    ) {
+      const [value, index, array] = this.mapForEachArgs(expr);
+
+      const next = expr.expr.expr;
+
+      // If we find array, stop recursing as array wouldn't be correct in the current map.
+      const [lastValue, list, lastRender] = this.divePropAccessMapOperations(
+        next,
+        `$${value}`,
+        !!array
+      );
+
+      return [
+        lastValue,
+        list,
+        () => {
+          lastRender();
```

Comment:
I'm confused how this function works. What does this callback do?

Reply:
uhh, yeah, I don't want to apply the `maps` to the VTL until all of them have been resolved. This function returns the variable name to put into the vtl `foreach`, so I want to find the deepest map, build the function to add the map logic to the vtl, return the variable to put in the foreach, and then execute (render) the maps into the `vtl` later. 

I would have done this differently if this class wasn't so stateful. haha

Again, will document.

Reply:
Why do you need a callback to be returned? Can you pass a flag into the recursive call to change how it renders instead of calling back?

Reply:
No, or at least I don't want to. This supports flattening maps for reduce and map/foreach (and maybe filter?), but this method is completely decoupled from the top level reduce/map logic including the `foreach` (map, reduce) and `initial value` (reduce) logic.

I didn't think of the function as a callback, but a closure to control when we render (similar to the body() method in reduce).